### ISSUE TITLE
project develop roadmap discuss

### ISSUE THREAD

**Issue by ywaby** [ywaby/godot-WaterPack#2]
## ripple
method
- texture blind
how to do？

## LOD
- godot buildin LOD
- script control mesh like [Zylann terrain](https://github.com/Zylann/godot_heightmap_native_plugin)

## bouyancy
- simple plane bouyancy
- wave bouyancy

method
- use add force at point
- use joint

## wave
now wave use “Gerstner Waves”, It not look reality.
“FFT” look  more better, I will try to add it.

## under water
feature
- volumetric light
- particle of something
- distort effect

method
- screen distort effect 
- fade with screen depth

**Comment by ywaby** [comment]
some reference documents, may be useful.

## reference
- Sea of Thieves（game）
- https://github.com/huwb/crest-oceanrender
- http://www.e2gamedev.com/water
- http://madebyevan.com/webgl-water/
- https://forum.unity.com/threads/ceto-scrawks-ocean.323545/page-3

wave gnerater
- FFT
    - https://www.keithlantz.net/2011/11/ocean-simulation-part-two-using-the-fast-fourier-transform/ 
- Gerstner Waves
    - https://80.lv/articles/tutorial-ocean-shader-with-gerstner-waves/
    - https://watersimulation.tumblr.com/post/115928482692/gerstner-waves 
    - https://developer.nvidia.com/gpugems/GPUGems/gpugems_ch01.html
    - https://opengllair.wordpress.com/2016/08/24/gerstner-wave-implementation/

infinite ocean
- https://github.com/huwb/crest-oceanrender
- https://github.com/SIsilicon/Godot-Ocean-Demo

lod
- godot terrain 

bouyancy
- https://github.com/dbrizov/Unity-WaterBuoyancy
- https://github.com/huwb/crest-oceanrender

reflection
- planar reflect
- Stencil Buffer

foam
- signed nearest distance
- https://80.lv/articles/tutorial-stylized-water-in-unreal/

ripple
- decal
- generate normalmap

underwater
- fade
- 体积光
- 泡泡等粒子

2d water
- unity Water 2D Tool
- https://github.com/valryon/water2d-unity
- http://rotatingcanvas.com/fragment-shader-to-simulate-water-surface-in-libgdx/


lowpoly water
- https://github.com/danielzeller/Lowpoly-Water-Unity

**Comment by SIsilicon** [comment]
I'm working on ripples. And it took a lot of time figuring out how it should be done.

## pass ripples data to shader

My first thought was to pass the position and size of each ripple to the shader via texture(since uniform arrays aren't supported yet). It worked kinda. The main drawback was the small amount of ripples you could make. See when making ripples from rain, you're going to need a lot more than just 25 ripples. At that point it would have chopped off 10 frames of performance. So I tried something else.

#ripple particle system

My second thought was to use a particle system of ripples. Each particle is a plane that has a ripple normal texture, and is placed _very_ close to the surface of the wet dirt(which is what I'm also making). Of course there's more to that. I also need to apply varying transparency to them according to the wetmap being used.
Now with this method I can have a *lot* of ripples. But then there is still the problem of blending them with the original(yet subtle) waves underneath. But can still be achieved. This is the method I'll be using.

**Comment by SIsilicon** [comment]
I also have [this][1] webpage on LOD that can also be referenced.

[1]:http://jeremybouny.fr/en/articles/webgl_lod/

**Comment by SIsilicon** [comment]
Good luck implementing FFT.
To do so requires iteration through render targets. And viewports are still kinda buggy for this job.

**Comment by ywaby** [comment]
I had also try to use ripple plane up water with use refraction ， but It seem like SCREEN_TEXTURE will not display water as water also use SCREEN_TEXTURE.
    idea is vertex shader use hightmap with wave,and fragment use normalmap with small wave.
ripple blend to normalmap with script. I think it better to get LOD and bouyancy too. something like Zylann terrain.

**Comment by ywaby** [comment]
Gerstner Waves  is repeat and smooth ,FFT look nice.
![](https://raw.githubusercontent.com/huwb/crest-oceanrender/master/img/teaser3.png)
FFT is not important, So not have priority.

**Comment by SIsilicon** [comment]
1:
FFT is usually, if not, always repeating as well.

2:
> I had also try to use ripple plane up water with use refraction ， but It seem like SCREEN_TEXTURE will not display water as water also use SCREEN_TEXTURE.
idea : vertex shader use hightmap with wave,and fragment use normalmap with small wave.
ripple blend to normalmap with script. I think it better to get LOD and bouyancy too. something like Zylann terrain.

It's pretty hard to understand what you are saying there. English isn't your language?

**Comment by SIsilicon** [comment]
Also the image you just showed me is made by gerstner waves.

**Comment by ywaby** [comment]
sorry for my poor English. 
I meat
1. script generate wave to hightmap
2. vertex shader use hightmap
3. mix normalmap in script
4. use normalmap in shader

- mix ripple normal to normalmap with script
- highmap to calc bouyancy
- only deal LOD range normalmap and highmap

**Comment by SIsilicon** [comment]
Yeah I was thinking the same thing after uploading it. I thought of making the ripples more procedural. It was also missing a complimentary splash from the rain drops(which still don't exist yet). I'll send another pull request with those features.

**Comment by ywaby** [comment]
add you as github Collaborators .

**Comment by SIsilicon** [comment]
Thanks! 😊

**Comment by SIsilicon** [comment]
[This](http://www.habrador.com/tutorials/unity-boat-tutorial/3-buoyancy/) should be a good reference on bouyancy.

**Comment by SIsilicon** [comment]
Progress update: I have created an infinite ocean using an almost traditional LOD based method, and not projected grid. However because of that there might be occasional seams popping up whenever the camera and or waves move. I'm still looking into how to mitigate that.
On another note the roles have also been updated. They look more like real rain drop ripples. Which also removed the need for a ripple normal texture as this is made procedurally in the shader.

**Comment by ywaby** [comment]
```
func set_noise_enabled(value):
	noise_enabled = value
	var old_noise_params = get_shader_param('noise_params', 0)
	old_noise_params.d = 1 if value else 0
	set_shader_param('noise_params', old_noise_params)
```
report bug
![screenshot_2018-06-11_10-40-15](https://user-images.githubusercontent.com/12198203/41210386-ea9a8616-6d63-11e8-9dee-87b4dab8f933.png)

and extends Spatial node for your infinite ocean  plz .

**Comment by ywaby** [comment]
Progress update:
1. add particle rain

2. underwater post process   
as depth texture unsupport in canvas_item shader. move it to godot 3.1 when godot supoort "cunstom post process".

3. buoyancy WIP
I will do plane buoyancy first.
add buoyancy points like https://wiki.unrealengine.com/Physical_Water_Surface

**Comment by SIsilicon** [comment]
Huh. I thought I got rid of that bug. Ok I should look into that.

**Comment by SIsilicon** [comment]
Hey by the way, I actually no a way we can create post process without Godot 3.1. I'll make a post process scene as a template you can work on. Good luck :)

**Comment by SIsilicon** [comment]
# Progress update
I have updated the ocean. The bug should be gone. I also put in some comments in its shader in case you want to change something.
I put a template for a post-process effect in the underwater shader (not underwater mesh shader). Now you have something too work off of. Here's how you use it.

- Use `SCREEN_UV` instead of `UV`.
- The mesh that must go with it must be generated by code. So the shader must be used by an `ImmediateGeometry` node. I'll uploaded the necessary script(it'll be inside of the script folder obviously).

Any questions?

**Comment by SIsilicon** [comment]
[progress update 1](https://github.com/ywaby/godot-WaterPack/commit/9c7d4d14b603add73bdac6f6653893def781389e)
[progress update 2](https://github.com/ywaby/godot-WaterPack/commit/7c1c7bc4fd46803e9e09b004a15c75a838b79407)

**Comment by SIsilicon** [comment]
Have you seen the new comments in the Godot devs forum? Someone is asking about cel shaded water. Should we add it? This _is_ a water pack after all.

**Comment by ywaby** [comment]
maybe later.

**Comment by ywaby** [comment]
finish base physic ，will merge with master tomorrow

**Comment by SIsilicon** [comment]
That's great! :D

**Comment by ywaby** [comment]
PS. reflection camera will conflict with shader default skymap reflecttion

**Comment by SIsilicon** [comment]
Which is why I was trying to figure out how to blend the two. I set the viewport to have a transparent background.
So when running the application is running where ever the reflection is transparent, the sky shows. And when opaque, only the planar reflection shows.

**Comment by ywaby** [comment]
if just tun off camera background will get a wrong reflection. the right way is godot add a api to tun off default skymap reflection.

**Comment by SIsilicon** [comment]
Will in the mean time, we'll just have to do a work around.
And if you turn off camera, then of course you'd get a wrong result. If you make that shader uniform `in_editor` set to true, then it won't take into account the camera based reflection. You should only set it to false when the camera is giving proper results. Which is why it is called `in_editor`. Because right now, the viewport texture isn't working in the editor. So enabling that uniform removed the reflection texture from the calculation. Is that understood? 😃

**Comment by SIsilicon** [comment]
The viewport's property `transparent_bg` is enabled so that it can blend with the default reflection. If you disable it (enable background in viewport), then the shade will have no idea, where there's geometry and where there's sky. And since the shaders lighting model automatically takes into account the skymap reflection, then it would be an opportunity to mix the global sky reflection with the local planar reflection.

**Comment by ywaby** [comment]
that is not good idea ,
main.gd load water.gd and water.gd load main.gd
```
plugin = get_node('/root/EditorNode/WaterPackPlugin')
```
why not update camera in main.gd
```
func forward_spatial_gui_input(p_camera, p_event):
	editor_camera = p_camera
	
	if p_camera.transform != editor_cam_backup_transform:
		for node in get_tree().get_nodes_in_group("water"):
			if node.use_reflection:
				var reflect_camera = node.get_node("./reflect_vp/reflect_cam")
				mirror(p_camera, reflect_camera, node) 
```

**Comment by SIsilicon** [comment]
1. `water.gd` isn't loading the plug-in. It is referencing it. So that we can get the camera from it. No need to work about cyclic reference here Ok?

2.  If we rely solely on the `main.gd` to supply the camera, then the reflections cannot work on runtime because the editor plugin only runs in the editor. Thus you wouldn't be getting any camera on runtime. So instead of doing that, we allow the `water.gd` to get the camera from the editor plugin *if* we need it.
In the editor? Use the camera supplied by the editor plugin.
Running the scene? Use the current camera in the scene. The editor plugin wouldn't exist on runtime anyway.

**Comment by ywaby** [comment]
your code is simple. I will remove my code

**Comment by ywaby** [comment]
but reflection is still not work in editor. because 
```
# water.gd
get_viewport().size # is return 0,0
```

**Comment by ywaby** [comment]
and you mention is [commithttps://github.com/ywaby/godot-WaterPack/commit/9c7d4d14b603add73bdac6f6653893def781389e)
```
 And have also changed the way foam is generated. Instead of being dependent on wave height, it's dependent on it's choppiness. 
```
where is it? water.shader not change at all?

**Comment by ywaby** [comment]
default reflection take at wrong place.
![screenshot_2018-06-15_14-52-15](https://user-images.githubusercontent.com/12198203/41454603-992b4fc6-70ac-11e8-969c-34f7d808f11b.png)
so the right solution is shutdown default reflection.
I had alseo try unshded type water, that will cause other problem.

**Comment by SIsilicon** [comment]
> but reflection is still not work in editor. 

This a general problem with the viewport. I'm looking for answers to get around that.
> and you mention is [commithttps://github.com/ywaby/godot-WaterPack/commit/9c7d4d14b603add73bdac6f6653893def781389e)

> And have also changed the way foam is generated. Instead of being dependent on wave height, it's dependent on it's choppiness. 
where is it? water.shader not change at all?

I was talking about my ocean shader. You shader looked so complex that I didn't really wanted to touch it.
> default reflection take at wrong place.

Are you sure that's not a graphics driver issue on your side? It looks a lot more than a wrong reflection I can see geometry sticking *outside* of the water(Just above the orange rock). How does that even happen.

**Comment by SIsilicon** [comment]
Just so you know I'm currently working on my own project. A volume renderer.
I'll copy this repository on Saturday along with Godot 3.0.3. and see to these.

**Comment by ywaby** [comment]
I don't think it a driver problem. the reason is default skymap reflection alway exist and at stone reflection mix with skymap reflection

**Comment by SIsilicon** [comment]
So what exactly is that 'thing' on top of the orange rock in the picture? That can't be a reflection, since the water geometry isn't in that region of the screen.

**Comment by ywaby** [comment]
that is my boat

**Comment by SIsilicon** [comment]
I see...
Ok then, like I said before.
> I'll copy this repository on Saturday along with Godot 3.0.3. and see to these.

**Comment by SIsilicon** [comment]
Ok I got the current repository and have a few things to say.

- I like what your thinking about how you'll be implementing the buoyancy force. I think it would be a good idea to allow the user to add the collision shapes and mesh instance them selves. Just like how you would with the other `PhysicsBodies`. I can take care of that.

- The reflections are working as expected for me(excluding the lack of clipping geometry)

- The toon water is not going according to the style of the water at Godot devs. Your supposed to just use a texture. And it should be unshaded.

**Comment by SIsilicon** [comment]
I'm polishing up the buoyancy for ya.
I'm going based off of [this][1].

[1]:(https://wiki.unrealengine.com/Physical_Water_Surface)

**Comment by ywaby** [comment]
I am work 2d buoyancy。 and improve it

**Comment by SIsilicon** [comment]
Ok then.
Also going to put in the toon water the people at Godot devs were talking about.

**Comment by SIsilicon** [comment]
Oh I also noticed that you were also working on 2D water. Now we have two. What to do...

Your water has reflections and more fleshed out waves.
My water has splashes and intuitive parameters.

What to do.. Should we combine them? And if so, who should?

**Comment by ywaby** [comment]
I will remove sprite 2d water and dev at poloygon 2d water

**Comment by SIsilicon** [comment]
Heeeyyy guess what?
I figured out how to remove water from inside the boat! The solution was simpler than I thought.
What I did was to create a boat hull. Which is just like the original boat, but without an inside. Then I made a `MeshInstance` with that hull. I have it a `ShaderMaterial` with a render priority of -1.
The shader itself was simple. `render_modes` were `unshaded` and `depth_draw_always`. And in the fragment shader, the `ALBEDO` was set to the `SCREEN_TEXTURE` with coordinates `SCREEN_UV`.

**Comment by ywaby** [comment]
good job,I also confuse with that problem. Now I try use area to auto detect water, and move bouyancy as a node type.

**Comment by SIsilicon** [comment]
> and move bouyancy as a node type.
That's what I did. :D

**Comment by SIsilicon** [comment]
Also, buoyancy is spelt with 'o' after 'u'.
b*uo*ancy.

**Comment by SIsilicon** [comment]
I put some "simple" functionality in the buoyancy's spatial gizmo. Now you can edit the points manually from the viewport too.

**Comment by SIsilicon** [comment]
Yeah that might work! The major problem would be ping ponging it though. Because a viewport can render only once per frame, you'd need a whole bunch of them to get this working, which would be a huge waste of memory. It's possible, but most likey impractical as of now. Maybe it will be better when Godot 4.0 comes around with a more flexible rendering pipeline.

Also, we kinda put this plug-in on a freeze. I don't know about ywaby, but I'm working on a game project right now. If you'd like, you can contribute to this project directly. I can handle any pull requests you might send. ;-)

**Comment by Lexpartizan** [comment]
`

> we kinda put this plug-in on a freeze.

`
Sorry to hear that, a very necessary plugin.
My knowledge was not enough even for the correct generation of textures h0 (k), h0 (-k). Therefore, I do not think that I can contribute to this project.
I just read your plans for creating FFT waves and thought you could tell me what I'm wrong about.
